### PR TITLE
Fixes #11297

### LINKED ISSUE #11297: [company prod issue] pipe file stream to process hangs
### What version of Bun is running?

1.1.9+bb13798d9

### What platform is your computer?

Darwin 23.1.0 arm64 arm

### What steps can reproduce the bug?

1) Ensure that you have `lz4` available locally (`sudo apt-get install lz4`  or `brew install lz4` or whatever)

2) Save this script in a `repro.js`

```javascript
const { spawn } = require("child_process");
const { createReadStream, mkdirSync } = require("fs");

// download a random archive to extract
downloadRandomArchive() 
  .then(async (archiveFile) => {

    // remove target directory & children
    await toPromise("rm dir", spawn("rm", ["-rf", ".target-dir"]));
    mkdirSync(".target-dir");

    // spawn processes
    const lz4 = spawn("lz4", ["-c", "-d", "-"]);
    const tar = spawn("tar", ["-x", "-C", ".target-dir"]);
    const lz4p = toPromise("lz4", lz4);
    const tarp = toPromise("tar", tar);

    // pipe lz4 to tar
    lz4.stdout.pipe(tar.stdin);

    // open the file to extract
    const f = createReadStream(archiveFile);

    // pipe the file to lz4
    f.pipe(lz4.stdin);

    // wait for both processes to finish
    await Promise.all([lz4p, tarp]);
  });

// ----- utilities -----

function downloadRandomArchive() {
  const target = "archive.tar.lz4";
  const dl = spawn(
    "curl",
    [
      "-L",
      "-r",
      "0-10000000", // only download a bit
      "https://huggingface.co/JCTN/fast-repo/resolve/main/repo.tar.lz4?download=true",
      "-o",
      target,
    ],
    {
      redirect: "manual",
    }
  );
  return toPromise("curl", dl).then(() => target);
}

function toPromise(name, p) {
  return new Promise((resolve, reject) => {
    p.on("exit", (code) =>  code ? reject(new Error(`${name} exited with code ${code}`)) : resolve());
  });
}
```

3) Check that it works in node: `node repro.js` will extract things in `.target-dir`  (will end with an error though - coz' the archive is not fully downloaded)

4) Check that `bun repro.js` hangs forever, without extracting anything

### What is the expected behavior?

Bun & Node should behave the same

### What do you see instead?

Bun hangs.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Fixes #11297

### How did you verify your code works?

There is a test

### CODE REVIEW

File: src/sys.zig [hunk 1]
```diff
@@ -2556,7 +2577,7 @@ pub fn linkatTmpfile(tmpfd: bun.FileDescriptor, dirfd: bun.FileDescriptor, name:
         if (Maybe(void).errnoSysFd(rc, .link, tmpfd)) |err| {
             switch (err.getErrno()) {
                 .INTR => continue,
-                .NOENT, .OPNOTSUPP, .PERM, .INVAL => {
+                .ISDIR, .NOENT, .OPNOTSUPP, .PERM, .INVAL => {
```
Comment (Jarred-Sumner) [inline src/sys.zig:2580]: This is unrelated

Forgot to include this earlier. It's for O_TMPFILE.

**Review by paperclover — APPROVED** [review]

**Review by gvilums — APPROVED** [review]
